<a href="https://colab.research.google.com/github/drameyjoshi/physics/blob/main/astro/course/what_are_stars.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [6]:
import astropy
import scipy
import numpy

In [7]:
G = astropy.constants.G.cgs                                 # Universal gravitational constant
c = astropy.constants.c.cgs                                 # Speed of light in vacuum
T = 365 * 24 * 60 * 60 * astropy.units.s                    # Number of seconds in a year
m_H = astropy.constants.m_p.cgs + astropy.constants.m_e.cgs # Mass of H atom in g
k_B = astropy.constants.k_B.cgs                             # Boltzmann constant in erg/K

# Estimation of physical parameters of the sun

## Mass

Assume that the earth moves in a circular orbit around the sun. Then the equation of motion is
\begin{equation}\tag{1}
\frac{mv^2}{r} = \frac{GmM_\odot}{r^2},
\end{equation}
where $m$ is the mass of the earth, $M$ is the mass of the sun, $v$ is the orbital speed of the earth, $r$ the radius of the orbit and $G$ is the universal gravitational constant. The speed of earth is
\begin{equation}\tag{2}
v = \frac{2\pi r}{T},
\end{equation}
where $T$ is the period of motion, 1 year. From equations (1) and (2) we get
\begin{equation}\tag{3}
M_\odot = 4\pi^2 \frac{r^3}{T^2G}.
\end{equation}
It takes $8$ minutes for light to reach the earth from the sun. This allows us to estimate $r$.

In [8]:
r = 8 * 60 * c * astropy.units.s # Radius of the earth's orbit.
M_s = 4 * scipy.constants.pi**2 * r**3 / (T**2 * G)
print(f"Mass of the sun is approximately {M_s}.")

Mass of the sun is approximately 1.772256672352156e+33 g.


We estimated the mass of sun to be $1.77 \times 10^{33}$ g. The actual mass is $1.90 \times 10^{33}$ g. Our estimate has an error of $(1.90 - 1.77)/1.90 ≈ 6.8\%$.

## Radius

The apparent size of the sun's disc at noon is $0.524^\circ$. If we denote the corresponding value in radians as $\theta_s$ then we have
\begin{equation}\tag{4}
\theta_\odot = \frac{0.524\pi}{180}.
\end{equation}
The radius of the sun is
\begin{equation}\tag{5}
R_\odot = \frac{r\theta_s}{2},
\end{equation}
where $r$ is the distance of the earth from the sun. We estimate is in the next cell.

In [9]:
R_s = 0.524 * scipy.constants.pi/180 * r/2
R_s_km = R_s.to("km")
print(f"Radius of the sun is approximately {R_s_km}.")

Radius of the sun is approximately 658022.2808465594 km.


The actual radius is $696,340$ km. Our estimate has an error of $5.5\%$.

## Mass density

We know the mass and the radius. The density is
\begin{equation}\tag{6}
\rho_\odot = \frac{3M_\odot}{4\pi R_\odot^3}.
\end{equation}

In [10]:
rho_s = 3 * M_s/(4 * scipy.constants.pi * R_s**3)
print(f"Density of the sun is approximately {rho_s}.")

Density of the sun is approximately 1.4849649846743935 g / cm3.


The actual value is $1.41$ g cm$^{-3}$. Our estimate has an error of close to $5 \%$.

## Surface temperature

The luminosity of sun is $L_\odot = 3.86 \times 10^{33}$ erg/s. Therefore, its radiance emittance is luminosity per unit surface area,
\begin{equation}\tag{7}
\mathcal{R}_\odot = \frac{L_\odot}{4\pi R_\odot^2}
\end{equation}
 Using Stefan's law, we can estimate the temperature to be
\begin{equation}\tag{8}
T_\odot = \left(\frac{\mathcal{R}_\odot}{\sigma}\right)^{1/4},
\end{equation}
where $\sigma$ is the Stefan-Boltzmann constant.

In [11]:
L_s = astropy.constants.L_sun.cgs # Solar luminosity in erg/s
Re_s = L_s/(4 * scipy.constants.pi * R_s**2)

# The package messed up with the units of Stefan-Boltzmann constant in cgs.
sigma = astropy.constants.sigma_sb.cgs.value * astropy.units.erg/(astropy.units.cm**2 * astropy.units.K**4 * astropy.units.s)

T_s = (Re_s/sigma)**(1/4)
print(f"Surface temperature of the sun is approximately {T_s}.")

Surface temperature of the sun is approximately 5934.952978938695 K.


Measured value of surface temperature is $5,772$ K. Our estimate has a $3\%$ error.

Thermal energy of a particle at temperature $T$ is
\begin{equation}\tag{9}
\mathcal{E}\text{(in erg)} = \frac{3}{2}k_BT,
\end{equation}
where $k_B$ is the Boltzmann constant. In electron volts it is
\begin{equation}\tag{10}
\mathcal{E} \text{(in eV)} = \frac{\mathcal{E}\text{(in erg)}}{1.6 \times 10^{-12}}
\end{equation}

In [12]:
E_surf = 1.5 * k_B * T_s
E_surf_eV = E_surf.to("eV")
print(f"Thermal energy of a particle at sun's sruface is {E_surf_eV}.")

Thermal energy of a particle at sun's sruface is 0.7671520157201405 eV.


This is way less than the ionization energy of atomic hydrogen. This is also far less than the bond dissociation energy of hydrogen molecule. Therefore, on the surface of the sun, hydrogen will exists in molecular form.

## Internal temperature

The internal temperature of the sun can be estimated using the virial theorem. However, to do that, we first need to compute the gravitational potential energy of the sun. If the mass density of sun is $\rho(r)$ then the potential energy of a shell of thickness $dr$ at a distance $r$ from the centre is

\begin{equation}\tag{11}
dU = -\frac{GM(r)}{r} dm = -\frac{GM(r)}{r} \rho(r) 4\pi r^2 dr.
\end{equation}

If we simplify the analysis by choosing $\rho = \rho_\odot$ throughout the sun's interior then we get
\begin{equation}\tag{12}
U = -\frac{3}{5}\frac{GM_\odot^2}{R_\odot}.
\end{equation}

According to the virial theorem, the kinetic energy is
\begin{equation}\tag{13}
K = -\frac{1}{2}U = \frac{3}{10}\frac{GM_\odot^2}{R_\odot}.
\end{equation}

If there are $N$ particles in the sun, then the kinetic energy is
\begin{equation}\tag{14}
K = \frac{3}{2}Nk_B T
\end{equation}

From equations (13) and (14) we get
\begin{equation}\tag{15}
T = \frac{GM_\odot^2}{5R_\odot Nk_B}.
\end{equation}

We know the values of all variables on the right hand side except $N$. We can estimate it from the solar mass assuming that all of it is hydrogen. If $m_H$ denotes the mass of the hydrogen atom then
\begin{equation}\tag{16}
Nm_H = M_⊙
\end{equation}

From (15) and (16) we get
\begin{equation}\tag{17}
T = \frac{Gm_HM_\odot}{5R_\odot k_B}.
\end{equation}


In [13]:
T = G * m_H * M_s/(5 * R_s * k_B)
print(f"Interior temperature of the sun is {T.value * 1E-6} million K.")

Interior temperature of the sun is 4.357854262452997 million K.


# Motion of a mass in a tunnel through the earth's centre

Let a mass $m$ be dropped in a tunnel that passes through the earth's centre and open at the antipodal position. At a distance $r$ from the centre, it experiences the gravitational attraction only of the mass enclosed by a sphere
of radius $r$, concentric with the earth. Thus, the equation of motion is
\begin{equation}\tag{18}
m\ddot{r} = -\frac{GM(r)m}{r^2}.
\end{equation}
Since
\begin{equation}\tag{19}
M(r) = \rho \frac{4\pi}{3}r^3,
\end{equation}
the equation of motion becomes
\begin{equation}\tag{20}
\ddot{r} = -\frac{4\pi}{3}\rho G r.
\end{equation}
The particle thus performs a simple harmonic motion with frequency,
\begin{equation}\tag{21}
\omega = \sqrt{\frac{4\pi}{3}\rho G}.
\end{equation}

In [14]:
M_e = astropy.constants.M_earth.cgs
R_e = astropy.constants.R_earth.cgs
V_e = 4 * scipy.constants.pi * R_e**3/3
rho_e = M_e/V_e

omega = numpy.sqrt(4*scipy.constants.pi * rho_e * G/3)
print(f"Angular frequency of oscillations of the particle is {omega}")

Angular frequency of oscillations of the particle is 0.0012394581825912545 1 / s


# Diffusion of a photon from the centre of the sun

The sun's centre has matter in plasma state. Electromagnetic radiation starting from the centre will be scattered by the charged particles of the plasma. It can be shown that the scattering is elastic in nature and has a cross section
\begin{equation}\tag{22}
\sigma_T = \frac{8\pi}{3} r_c^2,
\end{equation}
where
\begin{equation}\tag{23}
\frac{e^2}{r_c} = m_ec^2
\end{equation}
defines the classical electron radius $r_c$. If there are $n_e$ charged particles per unit volume then the mean free path is given by
\begin{equation}\tag{24}
l = \frac{1}{n_e\sigma_T}.
\end{equation}

In [22]:
e = astropy.constants.e.esu
r_c = e**2/(astropy.constants.m_e.cgs * c**2)
sigma_T = 8*scipy.constants.pi * r_c**2/3
print(f"The classical electron radius is {r_c.value} cm.")
print(f"Thomson scattering cross section is {sigma_T.value} cm^2.")

n_e = rho_s/m_H
l = 1/(n_e * sigma_T)
print(f"Mean free path is {l}.")

The classical electron radius is 2.817940324670788e-13 cm.
Thomson scattering cross section is 6.652458724930067e-25 cm^2.
Mean free path is 1.69408751926042 cm7 g2 / (Fr4 s4).


The mean free path of a photon after taking into account the variation of density of matter in the sun is $l ≈ 0.5$ cm. If a photon takes $N$ steps at random with a mean free path of $l$ then the distance it travels in $N$ steps is
\begin{equation}\tag{25}
D = \sqrt{N}\langle l \rangle.
\end{equation}
Therefore, the number of steps needed to cover a distance equal to the radius of the sun is
\begin{equation}\tag{26}
N_{es} = \frac{R_\odot^2}{\langle l \rangle^2}.
\end{equation}
Between two scattering events the photon travels at a speed $c$. The average between two such events is $\langle l \rangle/c$. Therefore, the time taken to cover $N_{es}$ is the time taken for a photon emerging at the centre to reach the surface of the sun is
\begin{equation}\tag{27}
\tau_{es} = N_{es}\frac{\langle l \rangle}{c}
\end{equation}
From (26) and (27) we get
\begin{equation}\tag{28}
\tau_{es} = \frac{R_\odot^2}{\langle l \rangle c}.
\end{equation}
Assuming that the correct $\langle l \rangle \approx 0.5$ cm, we estimate the time it takes for a photon starting at the centre of the sun to reach its surface.

In [26]:
tau_es = R_s**2/(0.5 * c)
tau_es_yr = tau_es.value/T
print(f"Time take for a photon at centre to escape the sun is {tau_es_yr.value} years.")

Time take for a photon at centre to escape the sun is 66285.38588560485 years.
